Where's my Bike feature

andrepeiry opened the pull request:
# *Where's my Bike ?* feature

Allow a user to store the location of its bike, to later find it easily in our app.

## Detailed content

- Add a small view model with an FSM-like behavior for ease of understanding and maintainability.
  - At first, there is no location / no use of the feature
  - Once the bike location is stored, the user can leave the application, and when (s)he comes back, the user can locate his/her bike by clicking the button, and the map will move to the location
  - Before going back to the *No location* state, it will ask confirmation to remove the stored location, as to avoid PEBCAK errors.
- A new button on the map for ease of use that transforms itself with the FSM above-mentioned.

## UI preview

The marker on the map
![image](https://github.com/user-attachments/assets/e93099fd-f259-4584-bdbe-501d443ea405)

## Future work

This PR does not work for a user that didn't use the phone GPS. This might be a good idea to improve this feature on this way, maybe with the crosshair of the "Add parking".

## Close

Close #322

andrepeiry committed:
fix: change default val in ProtoGenericSerializer

Having the defaultValue in the constructor brings the user to have a infinite constructor calls when trying to implements ProtoGenericSerializer.
Changing it to a function avoid this problem without much of convolution.

andrepeiry committed:
feat: add BikeLocationViewModel

Contains a BikeLocation, and a BikeLocationState for FSM-like behavior.

andrepeiry committed:
feat: add button for bike location storage

Add a second FAB on top of the "Recenter on location" one for ease of access.
Some things still needs to be debug (notably the camera movements) and small cleanup should be in order.
Maybe a refactor of the MapScreen with more modularity could be an idea for later.

andrepeiry committed:
refactor: improve zoomOnLocation

Improve it by setting behavioral modification on the map (tracking mode and recentering mode)

andrepeiry committed:
feat: bike location button UI

Add a button that follow the BikeLocationViewModel.

andrepeiry committed:
feat: draw BikeLocation on MapScreen

[BROKEN] The feature in this commit doesn't work. Will need fix later.

andrepeiry committed:
fix: draw BikeLocation on button click

[BROKEN] There are still two bugs that needs to be addressed:
- When parking the bike, the logo isn't drawn
- For each redraw, MapScreen is called 6 to 7 times, ruining the pop-up to forget the location

andrepeiry committed:
refactor: format with KTfmt

andrepeiry committed:
fix: change Setting for ProtoSerializer changes

andrepeiry committed:
fix: change MapViewModel onZoomLocation test

andrepeiry committed:
fix: change ParkingDetails goesToMapScreen test

andrepeiry committed:
fix: use new trigger for map redraw in MapScreen

andrepeiry committed:
feat: make navigationActions `?` in zoomOnLocation

andrepeiry committed:
fix: revert argument mix up during rebase

andrepeiry committed:
fix: graphical bug for bike location

Bug was between some logic of function call order, and other minutes details.
Clean the Log.d call.

andrepeiry committed:
fix: update zoomOnLocation usage from rebase

andrepeiry committed:
refactor: format with ktFmt

andrepeiry committed:
refactor: format with ktFmt

andrepeiry committed:
test: change Screen to Route in ZoneManager

andrepeiry committed:
refactor: revert a bad rebase

Remove Regex variables from #421 in MapScreen.

Pijukai commented on app/src/main/java/com/github/se/cyrcle/io/datastore/ProtoGenericSerializer.kt:
```diff
@@ -15,9 +15,12 @@ private const val SERIALIZE_TRUE = 1
+  override val defaultValue: T
+    get() = defaultValue()
+
+  abstract fun defaultValue(): T
```
Since it is a function with no override type maybe you could some documentation on it. A little one will do.

Pijukai commented on app/src/main/java/com/github/se/cyrcle/io/datastore/ProtoGenericSerializer.kt:
```diff
@@ -15,9 +15,12 @@ private const val SERIALIZE_TRUE = 1
+  abstract fun defaultValue(): T
 
   companion object {
     private fun writeNBytes(output: OutputStream, byteArray: ByteArray) {
```
Same here if possible.

Pijukai commented on app/src/main/java/com/github/se/cyrcle/model/map/BikeLocationViewModel.kt:
```diff
@@ -0,0 +1,177 @@
+   *   will be a valid location only if there was no IO error neither data corruption.
+   */
+  fun goToMyBike(): BikeLocation? {
+    Log.d("BikeLocationViewModel", "Fetching bike location from Preferences")
```
You left some Log, are they for debugging purpose or did you forget to remove them ?

andrepeiry committed:
feat: add custom marker for bike location

Pijukai commented on app/src/main/java/com/github/se/cyrcle/model/map/BikeLocationViewModel.kt:
```diff
@@ -0,0 +1,177 @@
+  private val latitudePreferenceKey = doublePreferencesKey("bike_location_latitude")
+
+  /**
+   * The state of where is the bike. At first, there is no location stored : [NO_LOCATION]. After
+   * parking the bike, the location is stored and the state is [LOCATION_STORED_UNUSED]. Once the
+   * user goes to the bike, the state is updated to [LOCATION_STORED_USED] and is ready for
+   * deletion.
+   */
+  enum class BikeLocationState {
+    NO_LOCATION,
+    LOCATION_STORED_UNUSED,
+    LOCATION_STORED_USED
+  }
+
+  private val _bikeLocationState = MutableStateFlow(NO_LOCATION)
+  val bikeLocationState = _bikeLocationState
+
+  /**
+   * A small class to contain the location of a bike, and the validity of said location. If
+   * [isValid] is false, the location is invalid and [location] is not guaranteed to be non-null.
+   *
+   * @property isValid Whether the location is valid
+   * @property location The location of the bike. Might be null if [isValid] is false
+   */
+  data class BikeLocation(val isValid: Boolean, val location: Point?) {
+    /**
+     * A small class to contain the location of a bike, and the validity of said location. If
+     * [isValid] is false, the location is invalid and [location] is not guaranteed to be non-null.
+     * This constructor is used to avoid calling [Point.fromLngLat] inside the code.
+     *
+     * @param isValid Whether the location is valid
+     * @param longitude The longitude of the bike
+     * @param latitude The latitude of the bike
+     */
```
Great documentation and use of an Enum. Makes the code easier to review !

Pijukai commented on app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:
```diff
@@ -938,3 +982,98 @@ fun PreviewCard(navigationActions: NavigationActions, parkingViewModel: ParkingV
+ * Composable function to display the Bike Location Floating action button. Will have different
+ * icons depending on the state of the bikeViewModel.
+ *
+ * @param bikeLocationViewModel The ViewModel for managing bike location-related data and actions.
+ * @param mapViewModel The ViewModel for managing map-related data and actions.
+ * @param navigationActions The actions to navigate to different screens.
+ */
+@Composable
+fun BikeLocationButton(
```
5 parameters but 3 documented seems like missing doc to me ^^.

andrepeiry replied:
Great eyes you have here !

Pijukai commented on app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:
```diff
@@ -938,3 +982,98 @@ fun PreviewCard(navigationActions: NavigationActions, parkingViewModel: ParkingV
+}
+
+@Composable
+fun RemoveBikeLocationPopup(
```
It was unclear since the beginning of the project but maybe you could doc here ? It is unblocking since we did not do it every time.

andrepeiry replied:
You know me, documentation is on the way ^^

sean-prz reviewed (commented):
Hey, something seems weird look at the following video : 
[Screen_recording_20241219_165325.webm](https://github.com/user-attachments/assets/0a97209b-fb8d-41d6-b0a4-7c02156c8639)

It seems like to me like this is a bug where the function to locate my bike is not working.

Pijukai commented:
> Hey, something seems weird look at the following video : [Screen_recording_20241219_165325.webm](https://github.com/user-attachments/assets/0a97209b-fb8d-41d6-b0a4-7c02156c8639)
> 
> It seems like to me like this is a bug where the function to locate my bike is not working.

I have the same problem with the recentering.

andrepeiry committed:
doc: add missing doc in ProtoGenericSerializer

For
- abstract fun defaultValue
- companion object > private fun writeNBytes

andrepeiry committed:
refactor: remove Log call in BikeLocationViewModel

andrepeiry committed:
doc: improve or complete MapScreen

For
- fun BikeLocationButton
- fun RemoveBikeLocationPopup

andrepeiry committed:
fix: map recentering on BikeLocation click

Add updateLocation in zoomOnLocation
Add redrawMarkers at MapScreen construction
Change argument in both zoomOnLocation in MapScreen

andrepeiry commented:
The latest commit (`fd88501` "fix: map recentering on BikeLocation click) should have fix the issue, I'll let you test again on your side !

sean-prz reviewed (commented):
I still have some major issue with this feature, take a look at the video 
[Screen_recording_20241219_210402.webm](https://github.com/user-attachments/assets/06bfbac1-ddde-4065-a48d-1263ae6f8168)

The pop up automatically closes itself.

**edit** I feel like there is a problem with your FSM not going back to a previous state when I scroll away from my parked bike ?

sean-prz reviewed (commented):
Don't you think it is weird to go from light blue to dark blue on the first click but to never have the inverse transition, to never be able to go to light blue.

I strongly feel like, to be consistent with the button below this new one, we should have a transition from dark blue to light blue when the user move is camera away from the parked bike. 

You could add this transition in `OnMoveListener`. 

To me it would feel way more intuitive.
That way the popup to unpark the bike only appears when clicking on the button while already being centered on the parked bike. 
And I could center myself on my parked biked multiple time sduring the same session without having the pop up

andrepeiry committed:
feat: improve bikeLocation button with movement

When moving the map right after clicking on the button in LOCATION_STORED_USED, the state of the FSM comes back to LOCATION_STORED_UNUSED.

andrepeiry commented:
To addressed @Pijukai and @sean-prz at the same time. The FSM is thought as "Once you got to your bike", you make it *ready to be deleted* state.

For the logic, I've taken in account your comments, and I've found an elegant way to merge them with the current FSM. Now, moving the map will turn the button light blue again.

As for the map reloading, I sadly have found no other way (excepting without creating 7 redraws of `MapScreen` and a lot of warnings and errors), so that is a necessary evil until a better solution could come to me in a dream.

sean-prz reviewed (changes requested):
Nice way better.

When I forget the parked bike the state is correctly expect that when I move the camera it changes back to a parked bike state, and then it **fatal crash** on click.
Let me show you in a video it will be easier : 

[Screen_recording_20241219_220719.webm](https://github.com/user-attachments/assets/1f093ab3-c70b-49fd-8a43-ee1cd59744f3)

andrepeiry committed:
fix: add logic for BikeLocationViewModel#mapMoved

sean-prz reviewed (approved):
The features  now fully works !!  🎉 
Good job on the reactivity for the bugs we spotted. 

Time for me to sign off, This is my last approve message, Adieu 🏳️ 🏳️

Pijukai reviewed (approved):
Well done André it works very well on my end as well ! It was not an easy feature but you succeeded in delivering and you kept improving it until our last comment So VERY WELL DONE ! This is also my last APPROVED!!!!!!!!!!

andrepeiry committed:
refactor: format with ktFmt
